app/models/manageiq/providers/azure/cloud_manager/metrics_capture.rb

```diff
@@ -77,76 +77,135 @@ def perf_collect_metrics(interval_name, start_time = nil, end_time = nil)
       return
     end
 
-    end_time   = end_time   ? end_time.utc   : Time.now.utc
-    start_time = start_time ? start_time.utc : (end_time - 4.hours) # 4 hours for symmetry with VIM
+    end_time       = end_time   ? end_time.to_time.utc   : Time.now.utc
+    start_time     = start_time ? start_time.to_time.utc : (end_time - 4.hours) # 4 hours for symmetry with VIM
+    time_interval  = start_time..end_time
+    counter_values = {}
 
     # This is just for consistency, to produce a :connect benchmark
     Benchmark.realtime_block(:connect) {}
 
     ems.with_provider_connection do |connection|
-      metrics_conn = Azure::Armrest::Insights::MetricsService.new(connection)
+      # TODO: code smell - should be in a 'collector' class instantiated here / find a better solution?
+      @connection = connection
```

Comment:
Hm making connection an instance var when it is in a block feels wrong, and it looks like it could be solved by just passing it in to the metrics_url/definition_url methods

Reply:
yes, passing shared variable in all the methods it is needed - similarly smelly approach. Other approach would be to create a class, where this variable will be passed on initialization, and it will be shared within this class as attribute. I didn't want to start such additional class making, because I see it as a part of the biggest refactoring I mentioned.